pyramid/predicates.py

```diff
@@ -137,7 +137,10 @@ def text(self):
     phash = text
 
     def __call__(self, context, request):
-        return self.val in request.accept
+        return any(
+            self.val == offer
+            for offer, q in request.accept.acceptable_offers([self.val])
```

Comment:
Since `q` is unused, would `_` be better?

Actually, for predicate purposes, should this just be `bool(request.accept.acceptable_offers([self.val])`?  Because `acceptable_values` does some normalization, etc.

Reply:
I think `bool(request.accept.acceptable_offers([self.val])` is a great idea. This PR was a late night attempt to get something out there to spur things along so I'm happy for the feedback before fixing the tests. :-)

Reply:
Currently according to the [documentation](https://github.com/Pylons/pyramid/blob/master/pyramid/config/views.py#L541) the value of the `accept` predicate can be media ranges (e.g. `text/*`, `*/*`) as well as media types, but `acceptable_offers()` only takes media types for its argument. I tried to lay out some of the issues [here](https://github.com/Pylons/pyramid/issues/1264#issuecomment-326554548) (point 1) and [here](https://github.com/Pylons/pyramid/issues/1264#issuecomment-329023914) (second half of comment) &mdash; the support for media ranges in the `accept` predicate was never working fully, as the old `NilAccept`'s (and so the old `MIMENilAccept`'s) [`__contains__`](https://github.com/Pylons/webob/blob/1.7-branch/webob/acceptparse.py#L198) called [`_check_offer`](https://github.com/Pylons/webob/blob/1.7-branch/webob/acceptparse.py#L340), which raised `ValueError("The application should offer specific types, got %r" % offer)`; and using media ranges for the `accept` predicate is ambiguous, and in my opinion does not make sense. If the predicate were to only take media types though, then `bool(request.accept.acceptable_offers([self.val])` would work well (...though only for a single media type. It could however be easily extended to take an iterable of media type strings &mdash; a finite number of media type offers makes sense I think.)

Reply:
Can you show a failing test (ideally a PR to pyramid's test suite but something standalone is fine too) that demonstrates what you're describing? Presumably a request with no accept header, and a view defined with a wildcard range?

Reply:
According to RFC 7231, e.g. `*/*` is (confusingly) *both* a media type and a media range. `acceptable_offers()` treats it as a media type (so the `*` has no special meaning), and a request with no accept header and a view defined with a wildcard range would match:

    >>> create_accept_header(header_value=None).acceptable_offers(['*/*'])
    [('*/*', 1.0)]

But

    >>> create_accept_header(header_value='text/*').acceptable_offers(['*/*'])
    []
    >>> create_accept_header(header_value='text/html').acceptable_offers(['*/*'])
    []

And 

    >>> create_accept_header(header_value=None).acceptable_offers(['text/*'])
    [('text/*', 1.0)]
    >>> create_accept_header(header_value='text/*').acceptable_offers(['text/*'])
    [('text/*', 1.0)]
    >>> create_accept_header(header_value='text/html').acceptable_offers(['text/*'])
    []

Working on branches and tests to help clarify what the behaviour was before, what has changed and why &mdash; should be done sometime this week.